rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,826 @@
+    /// 
+    /// The key is the `name()` provided by the `ModalDevTool` trait.
+    modal_dev_tools: HashMap<String, DevToolMetadata>,
+    /// The metadata for all registered dev commands.
+    /// 
+    /// The key is the `name()` provided by the `DevCommand` trait.
+    dev_commands: HashMap<String, DevCommandMetaData> 
+}
+```
+
+There are a few key operations that we will need to perform:
+
+```rust
+impl DevToolsRegistry {
+    /// Gets a reference to the specified modal dev tool by `ComponentId`.
+    fn get_tool_by_id(world: &World, component_id: ComponentId) -> Option<&dyn ModalDevTool> {
+        let resource = world.get_resource_by_id(component_id)?;
+        resource.downcast_ref()
+    }
+
+    /// Gets a mutable reference to the specified modal dev tool by `ComponentId`.
+    fn get_tool_mut_by_id(mut world: &mut World, component_id: ComponentId) -> Option<&mut dyn ModalDevTool> {
+        let resource = world.get_resource_mut_by_id(component_id)?;
+        resource.downcast_mut()
+    }
+
+    /// Gets the `DevCommandMetadata` for a given dev tool by name.
+    /// 
+    /// The supplied name should match the `DevCommand::name()` method.
+    fn get_command_metadata(name: &str) -> Option<&DevCommandMetadata> {
+        self.dev_commands.get(name)
+    }
+
+    /// Gets the `DevToolMetadata` for a given dev tool by name.
+    /// 
+    /// The supplied name should match the `DevCommand::name()` method.
+    fn get_tool_metadata(name: &str) -> Option<&DevToolMetadata> {
+        self.modal_dev_tools.get(name)
+    }
+
+    /// Looks up the `ComponentId` associated with the given name, as supplied by the `ModalDevTool` trait.
+    fn lookup_tool_component_id(&self, name: &str) -> Option<ComponentId> {
+       *self.get_tool_metadata(name)?.component_id
+    }
+
+    /// Gets a reference to the specified modal dev tool by type name, as supplied by the `ModalDevTool` trait.
+    fn get_tool_by_name(&self, world: &World, name: &str) -> Option<&dyn ModalDevTool> {
+        let component_id = self.lookup_tool_component_id(name)?;
+        DevToolsRegistry::get_by_id(world, component_id)
+    }
+
+    /// Gets a mutable reference to the specified modal dev tool by name.
+    fn get_tool_mut_by_name(&self, mut world: &mut World, component_id: ComponentId) -> Option<&mut dyn ModalDevTool> {
+        let component_id = self.lookup_tool_component_id(name)?;
+        DevToolsRegistry::get_mut_by_id(world, component_id)
+    }
+    
+    /// Iterates over the list of registered modal dev tools.
+    fn iter_tools(&self, world: &World) -> impl Iterator<Item = (ComponentId, &dyn ModalDevTool)> { 
+        self.modal_dev_tools.iter().map(|&id| (id, self.get(world, id)))
+    }
+
+    /// Mutably iterates over the list of registered modal dev tools.
+    fn iter_tools_mut(&self, mut world: &mut World) -> impl Iterator<Item = (ComponentId, &mut dyn ModalDevTool)> { 
+        self.modal_dev_tools.iter_mut().map(|&id| (id, self.get(world, id)))
+    }
```

Comment:
A cleanup is probably needed here.

We should probably have the methods match the impl above (see my comment above), only the name and return will change.